src/components/modules/dragNDrop.ts

```diff
@@ -89,28 +112,104 @@ export default class DragNDrop extends Module {
      * Try to set current block by drop target.
      * If drop target is not part of the Block, set last Block as current.
      */
-    const targetBlock = BlockManager.setCurrentBlockByChildNode(dropEvent.target as Node);
+    let targetBlock = BlockManager.setCurrentBlockByChildNode(dropEvent.target as Node);
 
     if (targetBlock) {
-      this.Editor.Caret.setToBlock(targetBlock, Caret.positions.END);
+      if (targetBlock.dropTarget === BlockDropZonePosition.Top) {
+        const currentIndex = BlockManager.getBlockIndex(targetBlock);
+
+        if (currentIndex > 0) {
+          targetBlock = BlockManager.getBlockByIndex(currentIndex - 1);
+          Caret.setToBlock(targetBlock, Caret.positions.END);
+        }
+        /**
+         *  @todo -Implement to drop a block before the first block,
+         * we should insert it before the first block.
+         */
+      } else {
+        Caret.setToBlock(targetBlock, Caret.positions.END);
+      }
     } else {
       const lastBlock = BlockManager.setCurrentBlockByChildNode(BlockManager.lastBlock.holder);
 
-      this.Editor.Caret.setToBlock(lastBlock, Caret.positions.END);
+      Caret.setToBlock(lastBlock, Caret.positions.END);
     }
 
+    BlockManager.clearDropTargets();
+    // Clear the selection.
+    BlockSelection.clearSelection();
+
     await Paste.processDataTransfer(dropEvent.dataTransfer, true);
   }
 
   /**
-   * Handle drag start event
+   * Process block drop event.
+   *
+   * @param dropEvent {DragEvent} - drop event
    */
-  private processDragStart(): void {
-    if (SelectionUtils.isAtEditor && !SelectionUtils.isCollapsed) {
-      this.isStartedAtEditor = true;
+  private processBlockDrop(dropEvent: DragEvent): void {
+    const { BlockManager } = this.Editor;
+
+    /**
+     * Remove drag image from DOM.
+     */
+    this.removeDragImage();
+
+    const selectedBlocks = BlockManager.blocks.filter(block => block.selected);
```

Comment:
I believe we have `BlockSelection.selectedBlocks` for that

Reply:
done

Reply:
done